File: old_trunk/cost/iona/orbacus/cpp/main/TestObjectWChar.cpp
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <includes.h>

#include "TestIntfWChar.h"
#include "TestMacro.h"

#include "TestObjectWChar.h"

using namespace CORBA;

// ----------------------------------------------------------------------
// Helper function for wcscmp. Should not be required
// with up-to-date compilers.
// ----------------------------------------------------------------------

int
mywcscmp(const wchar_t* ws1, const wchar_t* ws2)
{
  wchar_t wc1, wc2;
 
  do
  {
      wc1 = (wchar_t)*ws1++;
      wc2 = (wchar_t)*ws2++;
      if(!wc1)
          return wc1 - wc2;
  } 
  while(wc1 == wc2);

  return wc1 - wc2;
}

TestObjectWChar::TestObjectWChar(
    CORBA::ORB_ptr orb,
    ORBTest::Intf_ptr test_intf
)
    : TestObject(orb, test_intf)
{
}

TestObjectWChar::~TestObjectWChar()
{
}

bool
TestObjectWChar::is_supported(
    CORBA::Object_ptr obj
)
{
    ORBTest_WChar::Intf_var ti = (
	ORBTest_WChar::Intf::_narrow(obj)
    );
    return !CORBA::is_nil(ti);
}

void
TestObjectWChar::test_SII(
    CORBA::Object_ptr obj
)
{
    ORBTest_WChar::Intf_var ti = (
	ORBTest_WChar::Intf::_narrow(obj)
    );

    {
	WChar ret;
	ti->attrWChar(L'a');
	ret = ti->attrWChar();
	TEST(ret == L'a');
	ti->attrWChar(224);
	ret = ti->attrWChar();
	TEST(ret == 224);
	ti->attrWChar(0x20ac);
	ret = ti->attrWChar();
	TEST(ret == 0x20ac);

	WChar inOut, out;
	inOut = 1;
	ret = ti->opWChar(L'a', inOut, out);
	TEST(ret == L'b');
	TEST(inOut == L'b');
	TEST(out == L'b');
    }

    {
	WString_var ret;
	WString_var ws;

	ti->attrWString(L"");
	ret = ti->attrWString();
	ws = wstring_dup(L"");
	TEST(mywcscmp(ret, ws) == 0);

	ti->attrWString(L"Hello");
	ret = ti->attrWString();
	ws = wstring_dup(L"Hello");
	TEST(mywcscmp(ret, ws) == 0);

	WString_var inOut, out;
	inOut = wstring_dup(L"world!");
	ret = ti->opWString(L"Hello, ", inOut.inout(), out.out());
	ws = wstring_dup(L"Hello, world!");
	TEST(mywcscmp(ret, ws) == 0);
	TEST(mywcscmp(out, ws) == 0);
    }
}

void
TestObjectWChar::test_DII(
    CORBA::Object_ptr obj
)
{
    // REVISIT
}

File: cost/iona/orbacus/java/main/TestObjectWChar.java
// **********************************************************************
//
// Copyright (c) 2001
// IONA Technologies, Inc.
// Waltham, MA, USA
//
// All Rights Reserved
//
// **********************************************************************

package ORBTest;

import org.omg.CORBA.*;

public class TestObjectWChar
    extends test.common.TestBase
    implements TestObject
{
    private ORB m_orb;
    ORBTest.Intf m_test_intf;

    public
    TestObjectWChar(
        ORB orb,
	ORBTest.Intf test_intf
    )
    {
    	m_orb = orb;
	m_test_intf = test_intf;
    }

    public boolean
    is_supported(
	org.omg.CORBA.Object obj
    )
    {
	boolean is_supported = false;

	if (obj != null)
	{
	    try
	    {
		ORBTest_WChar.Intf ti = ORBTest_WChar.IntfHelper.narrow(obj);
		is_supported = true;
	    }
	    catch (BAD_PARAM e)
	    {
		is_supported = false;
	    }
	}

	return is_supported;
    }

    public void
    test_SII(
	org.omg.CORBA.Object obj
    )
    {
	ORBTest_WChar.Intf ti = ORBTest_WChar.IntfHelper.narrow(obj);

	{
	    char ret;
	    ti.attrWChar('a');
	    ret = ti.attrWChar();
	    TEST(ret == 'a');
	    ti.attrWChar((char)224);
	    ret = ti.attrWChar();
	    TEST(ret == (char)224);
	    ti.attrWChar((char)0x20ac);
	    ret = ti.attrWChar();
	    TEST(ret == (char)0x20ac);

	    CharHolder inOut = new CharHolder((char)1);
	    CharHolder out = new CharHolder();
	    ret = ti.opWChar('a', inOut, out);
	    TEST(ret == 'b');
	    TEST(inOut.value == 'b');
	    TEST(out.value == 'b');
	}

	{
	    String ret;

	    ti.attrWString("");
	    ret = ti.attrWString();
	    TEST(ret.equals(""));

	    ti.attrWString("Hello");
	    ret = ti.attrWString();
	    TEST(ret.equals("Hello"));

	    StringHolder inOut = new StringHolder("world!");
	    StringHolder out = new StringHolder();
	    ret = ti.opWString("Hello, ", inOut, out);
	    TEST(ret.equals("Hello, world!"));
	    TEST(out.value.equals("Hello, world!"));
	}

	{
	    CharHolder inOut = new CharHolder((char)1);
	    CharHolder out = new CharHolder();

	    try
	    {
		ti.opWCharEx('a', inOut, out);
		TEST(false);
	    }
	    catch (ORBTest_WChar.ExWChar ex)
	    {
		TEST(ex.value == 'b');
	    }
	}

	{
	    StringHolder inOut = new StringHolder("world!");
	    StringHolder out = new StringHolder();

	    try
	    {
		ti.opWStringEx("Hello, ", inOut, out);
		TEST(false);
	    }
	    catch (ORBTest_WChar.ExWString ex)
	    {
		TEST(ex.value.equals("Hello, world!"));
	    }
	}

    }

    public void
    test_DII(
	org.omg.CORBA.Object obj
    )
    {
	// REVISIT
    }
}
